File: live-build-config/config/includes.chroot/usr/local/bin/geoiplookup.pl
#!/usr/bin/perl -w
use Getopt::Long;
use Term::ANSIColor qw(:constants);

#命令行选项部分
my $help = "";
my $ip = "";
my $host = "";

local $Term::ANSIColor::AUTORESET = 1;

GetOptions(
'help|h'=>\$help,
'ip=s'=>\$ip,
'host=s'=>\$host
);


if($help){
print <<__HELP__;
Usage: $0 -ip 61.55.186.15 -host tanjiti.com

where:
-ip : Specify the ip address to query
-host: Specify the host address to query

__HELP__
exit 0;

}

chomp $ip;

die "You need to specify the ip address or hostname to query!!!! " if  ($ip eq "" and $host eq "");

#查询指定ip的地理信息
#query the specify ip geo information
geoiplookup($ip) if $ip ne "";

#查询指定host的地理信息
#query the specify host geo information
if ($host){
   my $ips=hosttoIP($host); 

   my @ips = split(' ',$ips);
   
   print BOLD RED "$host GEO information: \n";

   foreach (@ips){
      geoiplookup($_);
   }
}


#hosttoIP子函数：host反查获的域名对应的ip
sub hosttoIP{
   my $host = shift;
   my $ips = `host $host  | awk '{i=1; if(NF>0) do {if (\$i ~ /[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/) print \$i; i++;} while (i <=NF);}'`;
   return $ips;
}

#geoiplookup子函数,调用geoiplookup与对应的地理数据库获得指定ip的地理信息
sub  geoiplookup{
   my $ip = shift;
   chomp $ip;

   my $country = `geoiplookup -f /usr/share/GeoIP/GeoIP.dat $ip`;
   my $city = `geoiplookup -f /usr/share/GeoIP/GeoLiteCity.dat $ip`;
   my $ASNum = `geoiplookup -f /usr/share/GeoIP/GeoIPASNum.dat $ip`;

   print BOLD BLUE "$ip GEO information: \n";
   print "\t",$country;
   print "\t",$city;
   print "\t",$ASNum;
}

File: live-build-config/config/includes.chroot/usr/local/bin/geoiplookup.py
#!/usr/bin/env python

import sys

import maxminddb

import subprocess

import re

import json

import getopt


#specify the ip geo db path
GeoLit2_city = "/root/getipDB/GeoLite2-City.mmdb"


#ipInfo():获得指定ip的地理信息 ，以json格式返回

def ipInfo(ip):
readerCity = maxminddb.Reader(GeoLit2_city)

cityInfo = readerCity.get(ip)
print "\033[1;33;40m"
print "%s 's Geo information:" % ip
print "\033[0m"
print json.dumps(cityInfo, sort_keys=True, indent=2)

readerCity.close()


＃hostToIP():调用外部命令host，反查域名所在ip
def hostToIP(host):
   try:
      ipInfo = subprocess.check_output(['host',host],stderr=subprocess.STDOUT,)
   except subprocess.CalledProcessError as err:
      print "ERROR:",err
   else:
      ips = re.findall(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}',ipInfo)

return ips


＃hostInfo():获得指定域名的地理信息
def hostInfo(host):
   print "\033[1;31;40m"
   print "%s 's gep information: " % host
   print "\033[0m"
   ips = hostToIP(host)
   for item in ips:

ipInfo(item)chmod +x geoiplookup.py


＃usage(): 脚本帮助信息，显示各个选项的使用方法

def usage():

print "%s [-ip 61.55.186.15] [-host tanjiti.com]" % sys.argv[0]
print "\t -d or --domain : specify the host for query"
print "\t -i or --ip: specify the ip for query"
print "\t -h or --help: for more help"


＃main():主函数体

def main():
try:
opts,args = getopt.getopt(sys.argv[1:],"i:d:h",["ip=","domain=","help"])
except getopt.GetoptError as err:
print str(err)
usage()
sys.exit(2)
else:
for opt,val in opts:
if opt in ('-h','--help'):
usage()
sys.exit()
elif opt in ('-i','--ip'):
ipInfo(val)
elif opt in ('-d','--domain'):
hostInfo(val)
else:
assert False,"unhandled option"

if __name__ == "__main__":
main()
